File: 2023/CSharp/AocConsole/Day3.cs
using NaterLib;
namespace AocConsole;

public class Day3
{
    internal struct PartNumber
    {
        public PartNumber(int number, int start, int end)
        {
            this.number = number;
            this.start = start;
            this.end = end;
        }

        internal int number;
        internal int start;
        internal int end;
    }

    private static PartNumber ExtractPartNumber(string str, int pos)
    {
        if (!char.IsAsciiDigit(str[pos])) { return new PartNumber(0, pos, pos); }

        int step = pos;
        while (char.IsAsciiDigit(str[step])) { ++step; }
        int end = --step;
        int num = 0;
        int multiplier = 1;

        while (step >= 0 && char.IsAsciiDigit(str[step]))
        {
            num += int.Parse(str.Substring(step, 1)) * multiplier;
            multiplier *= 10;
            --step;
        }

        int start = ++step;
        return new PartNumber(num, start, end);
    }

    private static bool IsSymbol(char c)
    {
        return c != '.' && c != '\r' && !char.IsAsciiDigit(c);
    }

    private static bool IsGear(char c)
    {
        return c == '*';
    }

    private static bool IsPart(string[] lines, int row, int start, int end)
    {
        for (int i = start; i <= end; ++i)
        {
            if (row > 0)
            {
                if (IsSymbol(lines[row - 1][i])) { return true; }
                if (i > 0 && IsSymbol(lines[row - 1][i - 1])) { return true; }
                if (i < lines.Length - 1 && IsSymbol(lines[row - 1][i + 1])) { return true; }
            }

            if (row < lines.Length - 1)
            {
                if (IsSymbol(lines[row + 1][i])) { return true; }
                if (i > 0 && IsSymbol(lines[row + 1][i - 1])) { return true; }
                if (i < lines[row + 1].Length - 1 && IsSymbol(lines[row + 1][i + 1])) { return true; }
            }

            if (i > 0 && IsSymbol(lines[row][i - 1])) { return true; }
            if (i < lines[row].Length - 1 && IsSymbol(lines[row][i + 1])) { return true; }
        }

        return false;
    }

    private static List<PartNumber> WalkNeighbours(string[] lines, int row, int col)
    {
        List<PartNumber> numbers = [];

        // north
        if (row - 1 >= 0)
        {
            if (col > 0)
            {
                PartNumber nw = ExtractPartNumber(lines[row - 1], col - 1);
                if (nw.number != 0 && !numbers.Contains(nw)) { numbers.Add(nw); }
            }

            PartNumber n = ExtractPartNumber(lines[row - 1], col);
            if (n.number != 0 && !numbers.Contains(n)) { numbers.Add(n); }

            if (col < lines[row - 1].Length)
            {
                PartNumber ne = ExtractPartNumber(lines[row - 1], col + 1);
                if (ne.number != 0 && !numbers.Contains(ne)) { numbers.Add(ne); }
            }
        }

        // east and west
        if (col > 0)
        {
            PartNumber w = ExtractPartNumber(lines[row], col - 1);
            if (w.number !=0 && !numbers.Contains(w)) { numbers.Add(w); }
        }

        if (col < lines[row + 1].Length)
        {
            PartNumber e = ExtractPartNumber(lines[row], col + 1);
            if (e.number != 0 && !numbers.Contains(e)) { numbers.Add(e); }
        }

        // south
        if (row + 1 < lines.Length)
        {
            if (col > 0)
            {
                PartNumber sw = ExtractPartNumber(lines[row + 1], col - 1);
                if (sw.number != 0 && !numbers.Contains(sw)) { numbers.Add(sw); }
            }

            PartNumber s = ExtractPartNumber(lines[row + 1], col);
            if (s.number != 0 && !numbers.Contains(s)) { numbers.Add(s); }

            if (col < lines[row + 1].Length)
            {
                PartNumber se = ExtractPartNumber(lines[row + 1], col + 1);
                if (se.number != 0 && !numbers.Contains(se)) { numbers.Add(se); }
            }
        }

        return numbers;
    }

    public static int Part1(string data)
    {
        int sum = 0;
        
        string[] lines = data.Split('\n');
        for (int l = 0; l < lines.Length; ++l)
        {
            string line = lines[l];
            for (int i = 0; i < line.Length; ++i)
            {
                if (char.IsAsciiDigit(line[i]))
                {
                    PartNumber pn = ExtractPartNumber(line, i);
                    if (IsPart(lines, l, pn.start, pn.end))
                    {
                        sum += pn.number;
                    }
                    
                    i = pn.end + 1;
                }
            }
        }

        return sum;
    }

    public static int Part2(string data)
    {
        int sum = 0;

        string[] lines = data.Split('\n');
        for (int l = 0; l < lines.Length; ++l)
        {
            string line = lines[l];
            //Console.WriteLine($"{l,4}: {line}");
            for (int i = 0; i < line.Length; ++i)
            {
                if (IsGear(lines[l][i]))
                {
                    List<PartNumber> numbers = WalkNeighbours(lines, l, i);
                    if (numbers.Count == 2)
                    {
                        //Console.WriteLine($"\t{numbers.First().number} {numbers.Last().number}");
                        sum += numbers.First().number * numbers.Last().number;
                    }
                }
            }
        }
        return sum;
    }
}

File: 2023/Java/app/src/main/java/org/example/Day3.java
package org.example;

import java.util.ArrayList;
import java.nio.file.Path;
import java.util.List;

public class Day3 {
    private static class PositionalNumber {
        int start;
        int end;
        int number;

        static PositionalNumber extract(String str, int pos) {
            if (str == null || str.isEmpty() || pos < 0 || pos >= str.length() || !Character.isDigit(str.charAt(pos))) {
                return null;
            }

            int start = pos;
            while (start > -1 && Character.isDigit(str.charAt(start))) { --start; }
            ++start;

            int end = start;
            while (end < str.length() && Character.isDigit(str.charAt(end))) { ++end; } 

            if (end - start == 0) { 
                return null; }

            PositionalNumber pn = new PositionalNumber();
            pn.start = start;
            pn.end = end;
            pn.number = Integer.parseInt(str.substring(start, end));
            return pn;
        }

        static boolean containsMember(List<PositionalNumber> numbers, PositionalNumber number) {
            if (number == null) { return false; }
            for (PositionalNumber n : numbers) {
                if (number.start == n.start && n.end == number.end && n.number == number.number) { return true; }
            }

            return false;
        }
    }

    public static int part1(String fileName) {
        int sum = 0;
        
        List<String> lines = null;
        try {
            lines = java.nio.file.Files.readAllLines(Path.of(fileName));
        } catch (Exception e) {
            System.err.printf("Error loading file: %s\n", e.getMessage());
            return -1;
        }

        for (int y = 0; y < lines.size(); ++y) {
            for (int x = 0; x < lines.get(y).length(); ++x) {
                if (Character.isDigit(lines.get(y).charAt(x))) {
                    int s = x;
                    int n = parseNumber(lines.get(y), x);
                    int e = x + Integer.toString(n).length() - 1;

                    if (isPart(lines, y, s, e)) { sum += n;}
                    x = e + 1;
                }
            }
        }

        return sum;
    }

    public static int part2(String fileName) {
        int sum = 0;

        List<String> lines = null;
        try {
            lines = java.nio.file.Files.readAllLines((Path.of(fileName)));
        } catch (Exception e) {
            System.err.printf("Error loading file: %s\n", e.getMessage());
            return -1;
        }

        for (int y = 0; y < lines.size(); ++y) {
            String line = lines.get(y);
 //           System.out.printf("%s : %-4d\n", line, y);

            for (int x = 0; x < line.length(); ++x) {
                if (line.charAt(x) == '*') {
                    List<PositionalNumber> adjacentNumbers = getAdjacentUniqueNumbers(lines, y, x);
                    if (adjacentNumbers.size() == 2) { sum += adjacentNumbers.getFirst().number * adjacentNumbers.getLast().number; }
                }
            }
        }

        return sum;
    }

    /* TODO: Remove this tedium
     * We have to do lot's of bounds checks head for all 8 adjacent positions.
     * Can't we just just catch and ignore any out of bounds exception that might occur instead?
     */
    private static List<PositionalNumber> getAdjacentUniqueNumbers(List<String> lines, int gearY, int gearX) {
        List<PositionalNumber> numbers = new ArrayList<>();
        if (lines.size() == 0 || gearY < 0  || gearX < 0 || gearY >= lines.size()) { return numbers; }

        if (gearY > 0) {
            if (gearX > 0) {
                PositionalNumber nw = PositionalNumber.extract(lines.get(gearY - 1), gearX - 1);
                if (nw != null && !PositionalNumber.containsMember(numbers, nw)) { numbers.add(nw); }
            }

            if (gearX < lines.get(gearY).length() - 1) {
                PositionalNumber sw = PositionalNumber.extract(lines.get(gearY - 1), gearX + 1);
                if (sw != null && !PositionalNumber.containsMember(numbers, sw)) { numbers.add(sw); }
            }
        }

        if (gearX > 0) {
            PositionalNumber w = PositionalNumber.extract(lines.get(gearY), gearX - 1);
            if (w != null && !PositionalNumber.containsMember(numbers, w)) { numbers.add(w); }
        }

        if (gearX < lines.get(gearY).length() - 1) {
            PositionalNumber e = PositionalNumber.extract(lines.get(gearY), gearX + 1);
            if (e != null && !PositionalNumber.containsMember(numbers, e)) { numbers.add(e); }
        }

        if (gearY < (lines.size() - 1)) {
            if (gearX > 0) {
                PositionalNumber sw = PositionalNumber.extract(lines.get(gearY + 1), gearX - 1);
                if (sw != null && !PositionalNumber.containsMember(numbers, sw)) { numbers.add(sw); }
            }

            if (gearX < lines.get(gearY).length() - 1) {
                PositionalNumber se = PositionalNumber.extract(lines.get(gearY + 1), gearX + 1);
                if (se != null && !PositionalNumber.containsMember(numbers, se)) { numbers.add(se); }
            }
        }

        return numbers;
    }

    private static int parseNumber(String str, int pos) {
        int n = pos;
        while (n < str.length() && Character.isDigit(str.charAt(n))) { ++n; }
        return Integer.parseInt(str.substring(pos, n));
    }

    private static boolean isSymbol(char c) {
        return c != '.' && !Character.isDigit(c);
    }

    private static boolean isPart(List<String> lines, int line, int start, int end) {
        // north
        if (line > 0) {
            int n = start;
            if (n > 0) { n = n - 1; }
            while (n < lines.get(line - 1).length() && n <= end + 1) {
                if (isSymbol(lines.get(line - 1).charAt(n))) { return true; }
                ++n;
            }
        }

        // south
        if (line < lines.size() - 1) {
            int n = start;
            if (n > 0) { n = n - 1; }
            while (n < lines.get(line + 1).length() && n <= end + 1) {
                if (isSymbol(lines.get(line + 1).charAt(n))) { return true; }
                ++n;
            }
        }

        // west & east
        if (start > 0 && isSymbol(lines.get(line).charAt(start - 1))) { return true; }
        if (end < lines.get(line).length() - 1 && isSymbol((lines.get(line).charAt(end + 1)))) { return true; }

        return false;
    }
}
